crates/iroha_executor/src/default/isi/multisig/transaction.rs

```diff
@@ -0,0 +1,249 @@
+        let transaction_ttl_ms: u64 = host
+            .query_single(FindAccountMetadata::new(
+                multisig_account.clone(),
+                TRANSACTION_TTL_MS.parse().unwrap(),
+            ))
+            .dbg_unwrap()
+            .try_into_any()
+            .dbg_unwrap();
+        let instructions: Vec<InstructionBox> = host
+            .query_single(FindAccountMetadata::new(
+                multisig_account.clone(),
+                instructions_key(&instructions_hash),
+            ))?
+            .try_into_any()
+            .dbg_unwrap();
+        let proposed_at_ms: u64 = host
+            .query_single(FindAccountMetadata::new(
+                multisig_account.clone(),
+                proposed_at_ms_key(&instructions_hash),
+            ))
+            .dbg_unwrap()
+            .try_into_any()
+            .dbg_unwrap();
+        let now_ms: u64 = executor
+            .context()
+            .curr_block
+            .creation_time()
+            .as_millis()
+            .try_into()
+            .dbg_expect("shouldn't overflow within 584942417 years");
+        let mut approvals: BTreeSet<AccountId> = host
+            .query_single(FindAccountMetadata::new(
+                multisig_account.clone(),
+                approvals_key(&instructions_hash),
+            ))
+            .dbg_unwrap()
+            .try_into_any()
+            .dbg_unwrap();
+
+        approvals.insert(approver);
+
+        visit_seq!(executor.visit_set_account_key_value(&SetKeyValue::account(
+            multisig_account.clone(),
+            approvals_key(&instructions_hash),
+            Json::new(&approvals),
+        )));
+
+        let is_authenticated = quorum
+            <= signatories
+                .into_iter()
+                .filter(|(id, _)| approvals.contains(id))
+                .map(|(_, weight)| u16::from(weight))
+                .sum();
+
+        let is_expired = proposed_at_ms.saturating_add(transaction_ttl_ms) < now_ms;
+
+        if is_authenticated || is_expired {
+            // Cleanup the transaction entry
+            visit_seq!(
+                executor.visit_remove_account_key_value(&RemoveKeyValue::account(
+                    multisig_account.clone(),
+                    approvals_key(&instructions_hash),
+                ))
+            );
+
+            visit_seq!(
+                executor.visit_remove_account_key_value(&RemoveKeyValue::account(
+                    multisig_account.clone(),
+                    proposed_at_ms_key(&instructions_hash),
+                ))
+            );
+
+            visit_seq!(
+                executor.visit_remove_account_key_value(&RemoveKeyValue::account(
+                    multisig_account.clone(),
+                    instructions_key(&instructions_hash),
+                ))
+            );
+
+            if is_expired {
+                // TODO Notify that the proposal has expired, while returning Ok for the entry deletion to take effect
```

Comment:
Would using `Log` instruction work here? Or do you have some kind of client-retrievable notification?

Reply:
Good idea. Reflected every feedback in another PR #5240